README.md
<!--
---

title: "immugw"

custom_edit_url: https://github.com/codenotary/immudb/edit/master/README.md
---

-->

# immugw [![License](https://img.shields.io/github/license/codenotary/immudb)](LICENSE) <img align="right" src="img/Black%20logo%20-%20no%20background.png" height="47px" />

[![Build Status](https://travis-ci.com/codenotary/immudb.svg?branch=master)](https://travis-ci.com/codenotary/immudb)
[![Go Report Card](https://goreportcard.com/badge/github.com/codenotary/immugw)](https://goreportcard.com/report/github.com/codenotary/immugw)
[![Slack](https://img.shields.io/badge/join%20slack-%23immutability-brightgreen.svg)](https://slack.vchain.us/)
[![Discuss at immudb@googlegroups.com](https://img.shields.io/badge/discuss-immudb%40googlegroups.com-blue.svg)](https://groups.google.com/group/immudb)
[![Immudb Careers](https://img.shields.io/badge/careers-We%20are%20hiring!-blue?style=flat)](https://immudb.io/careers/)


**immugw** is the intelligent REST proxy that connects to immudb and provides a RESTful interface for applications. We recommend to run immudb and immugw on separate machines to enhance security


#### Build the binaries yourself

To build the binaries yourself, simply clone this repo and run

```
make all
```

#### immugw first start

##### Run immugw binary

```bash
# run immugw in the foreground
./immugw
```
##### Run immugw as a service

Service installation and management are supported on Linux, Windows, OSX and FreeBSD operating systems.

```
# install immugw service
./immuadmin service immugw install

# check current immugw service status
./immuadmin service immugw status

# stop immugw service
./immuadmin service immugw stop

# start immugw service
./immuadmin service immugw start
```

The linux service is using the following defaults:

| File or configuration   | location                   |
| ----------------------- | -------------------------- |
| executable              | /usr/sbin/immugw           |
| all configuration files | /etc/immugw                |
| pid file                | /var/lib/immugw/immugw.pid |
| log files               | /var/log/immugw            |

The FreeBSD service is using the following defaults:

| File or configuration   | location            |
| ----------------------- | ------------------- |
| executable              | /usr/sbin/immugw    |
| all configuration files | /etc/immugw         |
| pid file                | /var/run/immugw.pid |
| log files               | /var/log/immugw     |

The Windows service is using the following defaults:

| File or configuration   | location                             |
| ----------------------- | ------------------------------------
| executable              | Program Files\Immugw\immugw.exe      |
| configuration file      | ProgramData\Immugw\config\immugw.toml|
| all data files          | ProgramData\Immugw\                  |
| pid file                | ProgramData\Immugw\config\immugw.pid |
| log file                | ProgramData\Immugw\config\immugw.log |



Simply run `./immugw -d` to start immugw on the same machine as immudb (test or dev environment) or pointing to the remote immudb system ```./immugw --immudb-address "immudb-server"```.

If you want to stop immugw în that case you need to find the process `ps -ax | grep immugw` and then `kill -15 <pid>`. Windows PowerShell would be `Get-Process immugw* | Stop-Process`.

```bash
immu gateway: a smart REST proxy for immudb - the lightweight, high-speed immutable database for systems and applications.
It exposes all gRPC methods with a REST interface while wrapping all SAFE endpoints with a verification service.

Environment variables:
  IMMUGW_ADDRESS=0.0.0.0
  IMMUGW_PORT=3323
  IMMUGW_IMMUDB_ADDRESS=127.0.0.1
  IMMUGW_IMMUDB_PORT=3322
  IMMUGW_DIR=.
  IMMUGW_PIDFILE=
  IMMUGW_LOGFILE=
  IMMUGW_DETACHED=false
  IMMUGW_MTLS=false
  IMMUGW_SERVERNAME=localhost
  IMMUGW_AUDIT=false
  IMMUGW_AUDIT_INTERVAL=5m
  IMMUGW_AUDIT_USERNAME=immugwauditor
  IMMUGW_AUDIT_PASSWORD=
  IMMUGW_AUDIT_SIGNATURE=ignore
  IMMUGW_PKEY=
  IMMUGW_CERTIFICATE=
  IMMUGW_CLIENTCAS=

Usage:
  immugw [flags]
  immugw [command]

Available Commands:
  help        Help about any command
  version     Show the immugw version

Flags:
  -a, --address string            immugw host address (default "0.0.0.0")
      --audit                     enable audit mode (continuously fetches latest root from server, checks consistency against a local root and saves the latest root locally)
      --audit-interval duration   interval at which audit should run (default 5m0s)
      --audit-password string     immudb password used to login during audit; can be plain-text or base64 encoded (must be prefixed with 'enc:' if it is encoded)
      --audit-username string     immudb username used to login during audit (default "immugwauditor")
      --certificate string        server certificate file path (default "./tools/mtls/4_client/certs/localhost.cert.pem")
      --clientcas string          clients certificates list. Aka certificate authority (default "./tools/mtls/2_intermediate/certs/ca-chain.cert.pem")
      --config string             config file (default path are configs or $HOME. Default filename is immugw.toml)
  -d, --detached                  run immudb in background
      --dir string                program files folder (default ".")
  -h, --help                      help for immugw
  -k, --immudb-address string     immudb host address (default "127.0.0.1")
  -j, --immudb-port int           immudb port number (default 3322)
      --logfile string            log path with filename. E.g. /tmp/immugw/immugw.log
  -m, --mtls                      enable mutual tls
      --pidfile string            pid path with filename. E.g. /var/run/immugw.pid
      --pkey string               server private key path (default "./tools/mtls/4_client/private/localhost.key.pem")
  -p, --port int                  immugw port number (default 3323)
      --servername string         used to verify the hostname on the returned certificates (default "localhost")

Use "immugw [command] --help" for more information about a command.

```

### Docker

**immugw**  is also available as docker images on dockerhub.com.

| Component  | Container image                                |
| ---------- | ---------------------------------------------- |
| immugw     | https://hub.docker.com/r/codenotary/immugw     |

#### Run immugw

```
docker run -it -d -p 3323:3323 --name immugw --env IMMUGW_IMMUDB_ADDRESS=immudb codenotary/immugw:latest
```


#### Build the container images yourself

If you want to build the container images yourself, simply clone this repo and run

```
docker build -t myown/immugw:latest -f Dockerfile .
```

## Why immugw

**immugw** provides a simple solution to interact with immudb with REST protocol, without taking in charge the merkle tree root hash file management and concurrency related complexity.

#### immugw communication

**immugw**  proxies REST client communication and gRPC server interface. For security purposes immugw should not run on the same server as immudb. The following diagram shows how the communication works:

![immugw communication explained](img/immugw-diagram.png)

### API Documentation

You can find the swagger schema here:(https://github.com/codenotary/immugw/blob/master/swagger.json).  If you want to run the Swagger UI, simply run the following Docker command after you cloned the immugw repo, or the json file:

docker run -it -p 8081:8080 --name swagger-immugw -v ${PWD}/swagger.json:/openapi.json -e SWAGGER_JSON=/openapi.json  swaggerapi/swagger-ui

### CURL examples

#### Login
```shell script
curl --location --request POST '127.0.0.1:3323/login' \
--header 'Authorization;' \
--header 'Content-Type: application/json' \
--data-raw '{
    "user": "aW1tdWRi",
    "password": "aW1tdWRi"
}'
```
#### Use Database
```shell script
curl --location --request GET '127.0.0.1:3323/db/use/defaultdb' \
--header 'Content-Type: application/json' \
--header 'Authorization: {{token}}'
```
#### Login
```shell script
curl --location --request POST '127.0.0.1:3323/login' \
--header 'Authorization;' \
--header 'Content-Type: application/json' \
--data-raw '{
    "user": "aW1tdWRi",
    "password": "aW1tdWRi"
}'
```
#### Verified Set
```shell script
curl --location --request POST '127.0.0.1:3323/db/{database_name}/verified/set' \
--header 'Content-Type: application/json' \
--header 'Authorization: {{token}}' \
--data-raw '{
  "setRequest": {
    "KVs": [
      {
        "key": "a2V5MQ==",
	   "value": "dmFsMQ=="
      }
    ]
  }
}'
```
#### Verified Get
```shell script
curl --location --request POST '127.0.0.1:3323/db/{database_name}/verified/get' \
--header 'Content-Type: application/json' \
--header 'Authorization: {{token}}' \
--data-raw '{
  "keyRequest": {
    "key": "a2V5MQ=="
  }
}'
```
#### Verified Reference
```shell script
curl --location --request POST '127.0.0.1:3323/db/{database_name}/verified/setreference' \
--header 'Content-Type: application/json' \
--header 'Authorization: {{token}}' \
--data-raw '{
  "referenceRequest": {
    "key": "dGFnMQ==",
    "referencedKey": "a2V5MQ==",
    "atTx": "0"
  }
}'
```
#### Verified ZAdd
```shell script
curl --location --request POST '127.0.0.1:3323/db/{database_name}/verified/zadd' \
--header 'Content-Type: application/json' \
--header 'Authorization: {{token}}' \
--data-raw '{
  "zAddRequest": {
    "set": "c2V0MQ==",
    "score": 15.5,
    "key": "a2V5MQ==",
    "atTx": "0"
  }
}'
```
#### ZScan
```shell script
curl --location --request POST '127.0.0.1:3323/db/{database_name}/zscan' \
--header 'Content-Type: application/json' \
--header 'Authorization: {{token}}' \
--data-raw '{
  "set": "c2V0MQ=="
}'
```
#### History
```shell script
curl --location --request POST '127.0.0.1:3323/db/{database_name}/history' \
--header 'Authorization: {{token}}' \
--header 'Content-Type: application/json' \
--data-raw '{
  "key": "a2V5NQ=="
}'
```
#### Verified Transaction
```shell script
curl --location --request GET '127.0.0.1:3323/db/{database_name}/verified/tx/1' \
--header 'Content-Type: application/json' \
--header 'Authorization: {{token}}'
```
#### SQL Exec
```shell script
curl --location --request POST '127.0.0.1:3323/db/{database_name}/sqlexec' \
--header 'Authorization: {{token}}' \
--header 'Content-Type: application/json' \
--data-raw '{
    "sql":"CREATE TABLE mytable23 (id INTEGER, amount INTEGER, total INTEGER, title VARCHAR, content BLOB, isPresent BOOLEAN, PRIMARY KEY id)"
}'
```

#### SQL Exec insert
```shell script
curl --location --request POST '127.0.0.1:3323/db/{database_name}/sqlexec' \
--header 'Authorization: v2.public.eyJkYXRhYmFzZSI6IjUiLCJleHAiOiIyMDIxLTEwLTI4VDE4OjU1OjAyKzAyOjAwIiwic3ViIjoiaW1tdWRiIn3-aNUXqydajYFR9Aa7-q40JepLuA0tsPXeR1nRo75jA1H45RZZU9Twt6EVi-4bS4gpzeQcRNEdJs8U5oM5urcM.aW1tdWRi' \
--header 'Content-Type: application/json' \
--data-raw '{
    "sql":"INSERT INTO myTable23 (id, amount, title, content, isPresent) VALUES (2, 1000, '\''title 1'\'', x'\''626C6F6220636F6E74656E74'\'', true)"
}'
```
> byte arrays need to be hex encoded
#### SQL Query
```shell script
curl --location --request POST '127.0.0.1:3323/db/{database_name}/sqlquery' \
--header 'Authorization: {{token}}' \
--header 'Content-Type: application/json' \
--data-raw '{
    "sql":"SELECT * from mytable23;"
}'
```

#### SQL Verified sql row
Its possible also to tamperproof verify a SQL row.
```shell script
curl --location --request POST '127.0.0.1:3323/db/{database_name}/verified/row' \
--header 'Authorization: v2.public.eyJkYXRhYmFzZSI6IjUiLCJleHAiOiIyMDIxLTEwLTI4VDE4OjU1OjAyKzAyOjAwIiwic3ViIjoiaW1tdWRiIn3-aNUXqydajYFR9Aa7-q40JepLuA0tsPXeR1nRo75jA1H45RZZU9Twt6EVi-4bS4gpzeQcRNEdJs8U5oM5urcM.aW1tdWRi' \
--header 'Content-Type: application/json' \
--data-raw '{
    "row": {
            "columns": [
                "(testdb1.mytable23.id)",
                "(testdb1.mytable23.amount)",
                "(testdb1.mytable23.total)",
                "(testdb1.mytable23.title)",
                "(testdb1.mytable23.content)",
                "(testdb1.mytable23.ispresent)"
            ]
            },
            "values": [
                {
                    "n": "2"
                },
                {
                    "n": "1000"
                },
                {
                    "null": null
                },
                {
                     "s": "title 1"
                },
                {
                    "bs": "YmxvYiBjb250ZW50"
                },
                {
                    "b": true
                }
            ],
        "table": "mytable23",
        "pkValues": [
            {
                "n": "2"
            }
        ]
      }'
```
> byte arrays need to be b64 encoded
#### Logout
```shell script
curl --location --request POST '127.0.0.1:3323/logout' \
--header 'Authorization: {{token}}' \
--header 'Content-Type: application/json'
```


## License

immugw is [Apache v2.0 License](LICENSE).

immudb re-distributes other open-source tools and libraries - [Acknowledgements](https://github.com/codenotary/immudb/blob/master/ACKNOWLEDGEMENTS.md).


CHANGELOG.md
# CHANGELOG
All notable changes to this project will be documented in this file. This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
<a name="unreleased"></a>
## [Unreleased]


<a name="v1.3.1"></a>
## [v1.3.1] - 2023-07-27
### Bug Fixes
- **pkg/gw:** serverinfo endpoint is not bounded to any particular database ([#33](https://github.com/vchain-us/immudb/issues/33))

### Changes
- **README:** update swagger usage in README ([#30](https://github.com/vchain-us/immudb/issues/30))
- **pkg/gw:** support multiple kv pairs in set endpoint ([#34](https://github.com/vchain-us/immudb/issues/34))


<a name="v1.3.0"></a>
## [v1.3.0] - 2022-12-07

<a name="v1.3.0-RC1"></a>
## [v1.3.0-RC1] - 2022-12-05
### Bug Fixes
- **ci:** secure docker login command for password ([#24](https://github.com/vchain-us/immudb/issues/24))


<a name="v1.2.0"></a>
## [v1.2.0] - 2022-10-17
### Bug Fixes
- **ci:** update docker credentials

### Changes
- bump immudb version
- bump immudb version
- **ci:** update docker image and build prod images
- **pkg/api:** rename verifyRow request struct
- **pkg/api:** verified SQL get -> verified row renaming

### Features
- add verified sql get and fix tests
- **pkg/gw:** add immuerror server mux error handler


<a name="v1.0.5"></a>
## [v1.0.5] - 2021-08-25
### Changes
- bump immudb version
- add coverall
- remove codecov and fix go.mod
- immugw support immudb 0.9.x family
- upgrading for compatibility with immudb at 888ed37bf6cc
- **pkg/api:** clean obsolete swagger
- **pkg/gw:** handle corrupted data as 409 status conflict error


<a name="v0.9.2-RC1"></a>
## [v0.9.2-RC1] - 2021-04-26
### Changes
- add coverall
- remove codecov and fix go.mod
- immugw support immudb 0.9.x family
- upgrading for compatibility with immudb at 888ed37bf6cc
- **pkg/api:** clean obsolete swagger
- **pkg/gw:** handle corrupted data as 409 status conflict error


<a name="v0.8.1"></a>
## v0.8.1 - 2020-12-05
### Bug Fixes
- **cmd/immugw/command:** fix commandline_test error

### Changes
- remove verbose flag from test command

### Code Refactoring
- fixes signed root compatibility usage
- init immugw

### Features
- auditor verifies root signature
- **cmd/immugw:** add service management command


[Unreleased]: https://github.com/vchain-us/immudb/compare/v1.3.1...HEAD
[v1.3.1]: https://github.com/vchain-us/immudb/compare/v1.3.0...v1.3.1
[v1.3.0]: https://github.com/vchain-us/immudb/compare/v1.3.0-RC1...v1.3.0
[v1.3.0-RC1]: https://github.com/vchain-us/immudb/compare/v1.2.0...v1.3.0-RC1
[v1.2.0]: https://github.com/vchain-us/immudb/compare/v1.0.5...v1.2.0
[v1.0.5]: https://github.com/vchain-us/immudb/compare/v0.9.2-RC1...v1.0.5
[v0.9.2-RC1]: https://github.com/vchain-us/immudb/compare/v0.8.1...v0.9.2-RC1


cmd/immugw/command/cmd.go
/*
Copyright 2019-2020 vChain, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

	http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package immugw

import (
	"os"

	"github.com/codenotary/immudb/cmd/docs/man"
	c "github.com/codenotary/immudb/cmd/helper"
	"github.com/codenotary/immudb/cmd/version"
	"github.com/codenotary/immudb/embedded/logger"
	"github.com/codenotary/immudb/pkg/client"
	"github.com/codenotary/immudb/pkg/client/homedir"
	"github.com/codenotary/immudb/pkg/client/tokenservice"
	"github.com/codenotary/immugw/cmd/immugw/command/service"
	"github.com/codenotary/immugw/pkg/gw"
	"github.com/spf13/cobra"
	"github.com/spf13/viper"
	daem "github.com/takama/daemon"
)

func Execute() {
	version.App = "immugw"
	cl := Commandline{P: c.NewPlauncher(), config: c.Config{Name: "immugw"}}
	cmd, err := cl.NewCmd(gw.DefaultServer())
	if err != nil {
		c.QuitWithUserError(err)
	}
	if err = cmd.Execute(); err != nil {
		c.QuitWithUserError(err)
	}
}

// NewCmd creates a new immugw command
func (cl *Commandline) NewCmd(immugwServer gw.ImmuGw) (*cobra.Command, error) {
	cmd := &cobra.Command{
		Use:   "immugw",
		Short: "immu gateway: a smart REST proxy for immudb - the lightweight, high-speed immutable database for systems and applications",
		Long: `immu gateway: a smart REST proxy for immudb - the lightweight, high-speed immutable database for systems and applications.
It exposes all gRPC methods with a REST interface while wrapping all SAFE endpoints with a verification service.

Environment variables:
  IMMUGW_ADDRESS=0.0.0.0
  IMMUGW_PORT=3323
  IMMUGW_IMMUDB_ADDRESS=127.0.0.1
  IMMUGW_IMMUDB_PORT=3322
  IMMUGW_DIR=.
  IMMUGW_PIDFILE=
  IMMUGW_LOGFILE=
  IMMUGW_DETACHED=false
  IMMUGW_MTLS=false
  IMMUGW_SERVERNAME=localhost
  IMMUGW_AUDIT=false
  IMMUGW_AUDIT_INTERVAL=5m
  IMMUGW_AUDIT_USERNAME=immugwauditor
  IMMUGW_AUDIT_PASSWORD=
  IMMUGW_AUDIT_SIGNATURE=ignore
  IMMUGW_PKEY=./tools/mtls/4_client/private/localhost.key.pem
  IMMUGW_CERTIFICATE=./tools/mtls/4_client/certs/localhost.cert.pem
  IMMUGW_CLIENTCAS=./tools/mtls/2_intermediate/certs/ca-chain.cert.pem`,
		RunE:              cl.Immugw(immugwServer),
		PersistentPreRunE: cl.ConfigChain(nil),
	}

	cl.setupFlags(cmd, gw.DefaultOptions(), client.DefaultMTLsOptions())

	if err := viper.BindPFlags(cmd.Flags()); err != nil {
		return nil, err
	}

	setupDefaults(gw.DefaultOptions(), client.DefaultMTLsOptions())
	cmd.AddCommand(man.Generate(cmd, "immugw", "./cmd/docs/man/immugw"))

	cmd.AddCommand(version.VersionCmd())

	scl := service.NewCommandLine()
	scl.Register(cmd)

	return cmd, nil
}

// Immugw returns a new immudb gateway service
func (cl *Commandline) Immugw(immugwServer gw.ImmuGw) func(*cobra.Command, []string) error {
	return func(cmd *cobra.Command, args []string) (err error) {
		var options gw.Options
		if options, err = parseOptions(cmd); err != nil {
			return err
		}

		options = options.WithTokenService(tokenservice.NewFileTokenService().WithHds(homedir.NewHomedirService()))

		cliOpts := client.DefaultOptions().
			WithDir(options.Dir).
			WithPort(options.ImmudbPort).
			WithAddress(options.ImmudbAddress).
			WithHealthCheckRetries(1).
			WithMTLs(options.MTLs).
			WithMTLsOptions(options.MTLsOptions).
			WithMaxRecvMsgSize(4 * 1024 * 1024).
			WithAuth(true).
			WithConfig("")

		immuGwServer := immugwServer.WithOptions(options).WithCliOptions(*cliOpts)

		if options.Logfile != "" {
			if flogger, file, err := logger.NewFileLogger("immugw ", options.Logfile); err == nil {
				defer func() {
					if err = file.Close(); err != nil {
						c.QuitToStdErr(err)
					}
				}()
				immuGwServer.WithLogger(flogger)
			} else {
				return err
			}
		}
		plauncher := c.NewPlauncher()
		if options.Detached {
			if err := plauncher.Detached(); err == nil {
				os.Exit(0)
			}
		}

		var d daem.Daemon
		if d, err = daem.New("immugw", "immugw", "immugw"); err != nil {
			c.QuitToStdErr(err)
		}

		service := gw.Service{
			ImmuGwServer: immuGwServer,
		}

		d.Run(service)

		return
	}
}

func parseOptions(cmd *cobra.Command) (options gw.Options, err error) {
	dir, err := c.ResolvePath(viper.GetString("dir"), true)
	if err != nil {
		return options, err
	}
	port := viper.GetInt("port")
	address := viper.GetString("address")
	immudbport := viper.GetInt("immudb-port")
	immudbAddress := viper.GetString("immudb-address")
	audit := viper.GetBool("audit")
	auditInterval := viper.GetDuration("audit-interval")
	auditUsername := viper.GetString("audit-username")
	auditPassword := viper.GetString("audit-password")
	auditSignature := viper.GetString("audit-signature")
	pidfile, err := c.ResolvePath(viper.GetString("pidfile"), true)
	if err != nil {
		return options, err
	}
	logfile, err := c.ResolvePath(viper.GetString("logfile"), true)
	if err != nil {
		return options, err
	}
	mtls := viper.GetBool("mtls")
	detached := viper.GetBool("detached")
	servername := viper.GetString("servername")
	certificate, err := c.ResolvePath(viper.GetString("certificate"), true)
	if err != nil {
		return options, err
	}
	pkey, err := c.ResolvePath(viper.GetString("pkey"), true)
	if err != nil {
		return options, err
	}
	clientcas, err := c.ResolvePath(viper.GetString("clientcas"), true)
	if err != nil {
		return options, err
	}

	options = gw.DefaultOptions().
		WithDir(dir).
		WithPort(port).
		WithAddress(address).
		WithImmudbAddress(immudbAddress).
		WithImmudbPort(immudbport).
		WithAudit(audit).
		WithAuditInterval(auditInterval).
		WithAuditUsername(auditUsername).
		WithAuditPassword(auditPassword).
		WithAuditSignature(auditSignature).
		WithPidfile(pidfile).
		WithLogfile(logfile).
		WithMTLs(mtls).
		WithDetached(detached)
	if mtls {
		// todo https://golang.org/src/crypto/x509/root_linux.go
		options.MTLsOptions = client.DefaultMTLsOptions().
			WithServername(servername).
			WithCertificate(certificate).
			WithPkey(pkey).
			WithClientCAs(clientcas)
	}
	return options, nil
}

func (cl *Commandline) setupFlags(cmd *cobra.Command, options gw.Options, mtlsOptions client.MTLsOptions) {
	cmd.Flags().String("dir", options.Dir, "program files folder")
	cmd.Flags().IntP("port", "p", options.Port, "immugw port number")
	cmd.Flags().StringP("address", "a", options.Address, "immugw host address")
	cmd.Flags().IntP("immudb-port", "j", options.ImmudbPort, "immudb port number")
	cmd.Flags().StringP("immudb-address", "k", options.ImmudbAddress, "immudb host address")
	cmd.PersistentFlags().StringVar(&cl.config.CfgFn, "config", "", "config file (default path are configs or $HOME. Default filename is immugw.toml)")
	cmd.Flags().Bool("audit", options.Audit, "enable audit mode (continuously fetches latest root from server, checks consistency against a local root and saves the latest root locally)")
	cmd.Flags().Duration("audit-interval", options.AuditInterval, "interval at which audit should run")
	cmd.Flags().String("audit-username", options.AuditUsername, "immudb username used to login during audit")
	cmd.Flags().String("audit-password", options.AuditPassword, "immudb password used to login during audit; can be plain-text or base64 encoded (must be prefixed with 'enc:' if it is encoded)")
	cmd.Flags().String("audit-signature", "", "audit signature mode. ignore|validate. If 'ignore' is set auditor doesn't check for the root server signature. If 'validate' is set auditor verify that the root is signed properly by immudb server. Default value is 'ignore'")
	cmd.Flags().String("pidfile", options.Pidfile, "pid path with filename. E.g. /var/run/immugw.pid")
	cmd.Flags().String("logfile", options.Logfile, "log path with filename. E.g. /tmp/immugw/immugw.log")
	cmd.Flags().BoolP("mtls", "m", options.MTLs, "enable mutual tls")
	cmd.Flags().BoolP(c.DetachedFlag, c.DetachedShortFlag, options.Detached, "run immudb in background")
	cmd.Flags().String("servername", mtlsOptions.Servername, "used to verify the hostname on the returned certificates")
	cmd.Flags().String("certificate", mtlsOptions.Certificate, "server certificate file path")
	cmd.Flags().String("pkey", mtlsOptions.Pkey, "server private key path")
	cmd.Flags().String("clientcas", mtlsOptions.ClientCAs, "clients certificates list. Aka certificate authority")
}

func setupDefaults(options gw.Options, mtlsOptions client.MTLsOptions) {
	viper.SetDefault("dir", options.Dir)
	viper.SetDefault("port", options.Port)
	viper.SetDefault("address", options.Address)
	viper.SetDefault("immudb-port", options.ImmudbPort)
	viper.SetDefault("immudb-address", options.ImmudbAddress)
	viper.SetDefault("audit", options.Audit)
	viper.SetDefault("audit-interval", options.AuditInterval)
	viper.SetDefault("audit-username", options.AuditUsername)
	viper.SetDefault("audit-password", options.AuditPassword)
	viper.SetDefault("audit-signature", options.AuditSignature)
	viper.SetDefault("pidfile", options.Pidfile)
	viper.SetDefault("logfile", options.Logfile)
	viper.SetDefault("mtls", options.MTLs)
	viper.SetDefault("detached", options.Detached)
	viper.SetDefault("certificate", mtlsOptions.Certificate)
	viper.SetDefault("pkey", mtlsOptions.Pkey)
	viper.SetDefault("clientcas", mtlsOptions.ClientCAs)
}
